File: CSC_339_Concepts_of_Programming_Languages/ASSN3 FILES/assn3.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

/*
	Define the student data class. Methods are public, fields are private
*/
class Student {
	
	public:
		void setMajor(string m);
		void setLevel(string l);
		void setGPA(string g);
		void setCredits(string c);
		void setAdvisor(string a);
		string getID(void);
		string getName(void);
		string getMajor(void);
		string getLevel(void);
		string getGPA(void);
		string getCredits(void);
		string getAdvisor(void);
		void toString(void);
		Student();
		Student(string i, string n, string m, string l, string g, string c, string a);
	
	private:
		string ident;
		string name;
		string major;
		string level;
		string gpa;
		string credits;
		string advisor;
	
};

/*
	The constructors for Student.
*/
Student::Student(void) {}

Student::Student(string i, string n, string m, string l, string g, string c, string a){
	
	ident = i;
	name = n;
	major = m;
	level = l;
	gpa = g;
	credits = c;
	advisor = a;
	
}

/*
	The next several methods are simply getters and setters for the Student fields
*/
void Student::setMajor(string m) {
	major = m;
}

void Student::setLevel(string l) {
	level = l;
}

void Student::setGPA(string g) {
	gpa = g;
}

void Student::setCredits(string c) {
	credits = c;
}

void Student::setAdvisor(string a) {
	advisor = a;
}

string Student::getID(void) {
	return ident;
}

string Student::getName(void) {
	return name;
}

string Student::getMajor(void) {
	return major;
}

string Student::getLevel(void) {
	return level;
}

string Student::getGPA(void) {
	return gpa;
}

string Student::getCredits(void) {
	return credits;
}

string Student::getAdvisor(void) {
	return advisor;
}

void Student::toString(void) {
	cout << "\nID: " << ident << "\nName: " << name << "\nMajor: " << major << "\nLevel: " << level
		 << "\nGPA: " << gpa << "\nCredits: " << credits << "\nAdvisor: " << advisor << "\n";
}

/*
	Declare an array to hold students, as well as a counter to keep track
	of the current number of students
*/
Student students[10];
int stuCount = 0;

//Simply a method to print out all students
void printStudents(void) {
	for(int i = 0; i < stuCount; i++){
		students[i].toString();
		cout << "\n";
	}
}

/*
	Reads in student data from a .csv and populates it into an array
*/
void readIn(void) {
	ifstream records ("Student_Record.csv");
	string i, n, m, l, g, c, a;
	while(records.good()) {
		getline(records, i, ',');
		getline(records, n, ',');
		getline(records, m, ',');
		getline(records, l, ',');
		getline(records, g, ',');
		getline(records, c, ',');
		getline(records, a, '\n');
		
		Student stu(i, n, m, l, g, c, a);
		students[stuCount] = stu;
		stuCount++;
	}
	records.close();
}

/*
	Allows the user to create a new student
*/
void createStudent(void) {
	cin.ignore();
	string i, n, m, l, g, c, a;
	cout << "Input 9 digit ID number: ";
	getline(cin,i);
	cout << "Input name: ";
	getline(cin,n);
	cout << "Input major: ";
	getline(cin,m);
	cout << "Input level: ";
	getline(cin,l);
	cout << "Input GPA: ";
	getline(cin,g);
	cout << "Input credits: ";
	getline(cin,c);
	cout << "Input advisor: ";
	getline(cin,a);
	
	Student stu(i, n, m, l, g, c, a);
	students[stuCount] = stu;
	stuCount++;
}

/*
	Allows the user to find a specific student by ID number
*/
int getStudent(void){
	cin.ignore();
	string idnum;
	cout << "Input nine digit student ID: ";
	getline(cin, idnum);
	for(int i = 0; i < stuCount; i++){
		if(students[i].getID() == idnum)
			return i;
	}
	cout << "Student not found\n";
	return -1;
}

/*
	Allow a user to modify student info
*/
void modifyStudent(void){
	int pos;
	string m,l,g,c,a;
	pos = getStudent();
	if(pos == -1)
		return;
	cout << "Current student info:";
	students[pos].toString();
	cout << "Input new major: ";
	getline(cin, m);
	students[pos].setMajor(m);
	cout << "Input new level: ";
	getline(cin, l);
	students[pos].setLevel(l);
	cout << "Input new GPA: ";
	getline(cin, g);
	students[pos].setGPA(g);
	cout << "Input new credits: ";
	getline(cin, c);
	students[pos].setCredits(c);
	cout << "Input new advisor: ";
	getline(cin, a);
	students[pos].setAdvisor(a);
}

/*
	A method to sort students by ID number
	I considered using quick sort but in the end went with bubble sort
	mainly because its a lot easier to implement and I'm not overly
	worried about efficiency in this program
*/
void idSort(void){
	for(int i = 1; i < stuCount; i++){
		for(int j = 0; j < stuCount-1; j++){
			if(students[j].getID().compare(students[j+1].getID()) > 0){
				Student temp = students[j];
				students[j] = students[j+1];
				students[j+1] = temp;
			}
		}
	}
}

/*
	Sorts by student name
*/
void nameSort(void){
	for(int i = 1; i < stuCount; i++){
		for(int j = 0; j < stuCount-1; j++){
			if(students[j].getName().compare(students[j+1].getName()) > 0){
				Student temp = students[j];
				students[j] = students[j+1];
				students[j+1] = temp;
			}
		}
	}
}

/*
	Sorts by student GPA
*/
void gpaSort(void){
	for(int i = 1; i < stuCount; i++){
		for(int j = 0; j < stuCount-1; j++){
			if(students[j].getGPA().compare(students[j+1].getGPA()) < 0){
				Student temp = students[j];
				students[j] = students[j+1];
				students[j+1] = temp;
			}
		}
	}
}

/*
	A menu of available options for the user
*/
void showOptions(void) {
	cout << "\nSelect Option";
	cout << "\n1 - Enter student information";
	cout << "\n2 - Modify student information";
	cout << "\n3 - Lookup student by ID";
	
	cout << "\n4 - Display all students";
	cout << "\n5 - Display all students by Name";
	cout << "\n6 - Display all students by GPA";
	
	cout << "\n7 - Exit\n";
}

int main(){
	//Reads in data from a .csv file
	readIn();
	
	int choice = 0;
	
	//Show the options to the user, and continue to display the menu
	//untill the 'Exit' option is selected
	while(choice != 7){
		showOptions();
		cin >> choice;
		switch(choice){
			case 1:
				createStudent();
				break;
			case 2:
				modifyStudent();
				break;
			case 3:
				int pos;
				pos = getStudent();
				if(pos != -1) {
					students[pos].toString();
				}
				break;
			case 4:
				idSort();
				printStudents();
				break;
			case 5:
				nameSort();
				printStudents();
				break;
			case 6:
				gpaSort();
				printStudents();
				break;
		}
	}	
	return 0;
}

File: CSC_339_Concepts_of_Programming_Languages/ASSN3 FILES/assn3.py
import csv

"""
Define a structure for the student data class. Has one constructor and a toString method
"""
class Student:
	def __init__(self, ident, name, major, level, gpa, credit, advisor):
        	self.ident = ident
	        self.name = name
	        self.major = major
	        self.level = level
	        self.gpa = gpa
	        self.credits = credit
	        self.advisor = advisor
	
	def __str__(self):
		return "\nID: " + self.ident + "\nName: " + self.name + "\nMajor: " + self.major + "\nLevel: " + self.level + "\nGPA: " + self.gpa + "\nCredits: " + self.credits + "\nAdvisor: " + self.advisor

#Create a list to hold Students
students = []

#A method to print all Students
def printStudents(stu):
	for x in stu:
		print x

#A lambda expression to sort students by ID
def sortID():
	printStudents(sorted(students, key=lambda student: student.ident))

#A method to sort by name
def sortName():
	printStudents(sorted(students, key=lambda student: student.name))

#A method to sort by GPA
def sortGPA():
	printStudents(sorted(students, key=lambda student: student.gpa))

#Allow a user to create a new student and add it to the student list
def createStudent():
	i = raw_input("Enter nine digit ID number: ")
	n = raw_input("Enter name: ")
	m = raw_input("Enter major: ")
	l = raw_input("Enter level: ")
	g = raw_input("Enter GPA: ")
	c = raw_input("Enter credits: ")
	a = raw_input("Enter advisor: ")
	st = Student(i,n,m,l,g,c,a,)
	students.append(st)

#Lookup a student by ID number	
def lookupStudent():
	i = raw_input("Please enter nine digit ID number: ")
	for x in students:
		if i == x.ident:
			return x
	return "Student not found"

#Allow a user to modify a students info
def modifyStudent():
	stu = lookupStudent()
	print "Current student info:"
	print stu
	stu.major = raw_input("Enter new major: ")
	stu.level = raw_input("Enter new level: ")
	stu.gpa = raw_input("Enter new GPA: ")
	stu.credits = raw_input("Enter new credits: ")
	stu.advisor = raw_input("Enter new advisor: ")
	
#Read in data from a .csv file and populate the Student list	
with open('Student_Record.csv', 'r') as csvfile:
	recordReader = csv.reader(csvfile)
	for row in recordReader:
		stu = Student(row[0], row[1], row[2], row[3], row[4], row[5], row[6])
		students.append(stu)

#Show available options to the user
def showOptions():
	print "\n1 - Enter student information"
	print "2 - Modify student information"
	print "3 - Lookup student by ID"
	print "4 - Display all students"
	print "5 - Display students by name"
	print "6 - Display students by GPA"
	print "7 - Exit"

#Ask the user to select an option, and return to this screen
#untill the 'Exit' option is selected
choice = 0
while int(choice) <> 7:
	showOptions()
	choice = raw_input("Select option: ")
	if choice == '1':
		createStudent()
	elif choice == '2':
		modifyStudent()
	elif choice == '3':
		print lookupStudent()
	elif choice == '4':
		sortID()
	elif choice == '5':
		sortName()
	elif choice == '6':
		sortGPA()
